### ISSUE TITLE
Plug in display dimming

### ISSUE THREAD

**Issue by nigelbreretonthelegend-sys** [Anezium/Rokid-Nexus#37]
Subject: Request for supported screen-dark microphone capture in Nexus plugins

Hello Nexus team,

We tested a private, ordinary phone-side Nexus plugin on Rokid RV101 glasses using the public bus-client SDK 0.17.0 and Nexus hubs 1.4.8. It acquires the glasses microphone through the approved `microphone` capability and receives raw 16 kHz mono PCM16LE frames with sequence and timing data.

The remaining obstacle is display behavior. Keeping the plugin’s glasses surface open leaves the HUD lit, even when the plugin stops sending live counter updates. Hiding its last surface closes the plugin session under the current plugin lifecycle. Our ADB diagnostics associated the active Nexus glasses-hub window with a screen-bright wake lock.

As a diagnostic only, we sent Android’s sleep key event to the glasses during capture. The display went dark, while microphone frames continued until a deliberate Stop after 5:08: 15,410 frames, sequence 0–15409, zero missing, duplicate, out-of-order or backwards-timestamp events, and a 340 ms maximum inter-frame interval. Stop woke the display. This demonstrates that the audio path can operate with the display dark; it does not establish a supported plugin method or a battery-saving figure.

Is there already a supported way for an approved microphone plugin to retain its lease and phone-side controls while allowing the glasses display to sleep normally, without ADB? If not, would you consider an opt-in sleep-eligible or audio-only plugin mode that preserves capability approval, lease revocation, and normal user control? We would welcome your view on the smallest appropriate change and whether it fits the Nexus roadmap.

Can a microphone lease remain active with only a live activity or pin visible, while the full surface is closed and normal display sleep is allowed?

This is a feasibility request, not a request to replace or debug the installed hubs.

**Comment by Anezium** [comment]
> [!NOTE]
> 🤖 **GPT-6 Astra responding on my behalf**

Yes — this is now supported in **Nexus 1.4.10** with **SDK 0.18.0** (plugin API remains version 3).

After your `NexusAudioCallbacks.onAudioStarted` callback, call `surface.detach()` (or `surface.hide(detach = true)`) on your last ordinary surface. The hub accepts this only when your plugin owns the active microphone lease and the stream has started. It removes the surface while keeping the audio session and SDK foreground service alive. No ADB or screen-control permission is needed by the plugin.

- `onNexusBackground()` lets you release your UI references while retaining your `NexusAudioSession`.
- Opening the plugin again delivers `onNexusOpen(PluginOpenTypes.RESUME)`, so you can show microphone controls again.
- The phone hub shows **Listening in the background** with a **Stop** button. Lease release/revocation, link loss, or plugin process death ends the background session.
- A pin can remain without waking or holding the display on. For a live microphone indicator, use a short TTL renewed by incoming audio; the Sample uses 5 seconds, renewed every 2 seconds, so a crash cannot leave a stale indicator.

This makes capture eligible for normal display sleep; it does not force sleep or override the firmware timeout, and an open launcher or another foreground screen can still keep the display awake.

We tested this on a Samsung phone and Rokid glasses: continuing PCM delivery, launcher/phone resume, back-to-detach, Stop from the phone while the glasses were asleep, Bluetooth link loss, and plugin process death. For the idle-sleep test we temporarily set the device timeout to 15 seconds and restored it afterwards. We have not measured battery savings or repeated your five-minute frame-integrity benchmark.

Update both hubs and rebuild your plugin with:

```kotlin
implementation("com.github.Anezium.Rokid-Nexus:bus-client:sdk-v0.18.0")
```

[Hubs 1.4.10](https://github.com/Anezium/Rokid-Nexus/releases/tag/v1.4.10) · [SDK 0.18.0](https://github.com/Anezium/Rokid-Nexus/releases/tag/sdk-v0.18.0) · [SDK lifecycle example](https://github.com/Anezium/Rokid-Nexus/blob/sdk-v0.18.0/docs/PLUGIN_SDK.md#let-the-display-sleep-while-listening) · [Sample implementation](https://github.com/Anezium/Rokid-Nexus/tree/sample-v1.0.3/plugins/sample)

Your existing approval and microphone capability still apply. On older hubs, `detach` follows the ordinary hide/close behavior, so update the hubs before trying it. Please try it with your plugin and tell us whether it meets your screen-dark capture use case.

**Comment by nigelbreretonthelegend-sys** [comment]
Thanks — 1.4.10/SDK 0.18.0 sounds ideal for the screen-dark listening use case. We’ll test it once our current baseline testing is complete.
While you’re looking at the audio side, could I ask about microphone scene selection on the Rokid glasses?
The glasses have a four-microphone array and Rokid appears to support different microphone processing scenes — broadly near-field/wearer focused, far-field/forward focused, and wider/omnidirectional capture.
With the current Nexus audio lease we receive the processed 16 kHz mono PCM stream, but I can’t see a public plugin API for selecting the upstream microphone scene.
Is microphone/audio scene selection available through the CXR interface used by Nexus, or could Nexus expose it safely to plugins?
We have encountered Rokid APIs elsewhere such as changeAudioSceneId(...) / switchMicScene(...), but the scene IDs and APIs appear to differ between SDK families, so I don’t want to assume they apply to the RV101.
The intended Jeeves use case is to switch deliberately between listening primarily to the wearer, listening to somebody in front of the wearer, and capturing a wider conversation/environment, while retaining the existing Nexus audio lease.
If the RV101/CXR interface supports this, knowing the supported method and valid scene values would be extremely useful.

**Comment by Anezium** [comment]
Hey, so nope, and it's not something I can add on my side unfortunately.

Nexus gets the glasses mic through Rokid's external CXR-L interface, and that interface only has start/stop audio with a codec. 
The processing you're hearing (the near-field wearer beamformer on the 4 mics) is chosen inside Rokid's own service on the glasses, and the omni mode only exists behind their translation feature, not on the audio path apps can reach. 
The changeAudioSceneId / switchMicScene calls you've seen come from other Rokid SDKs and don't apply to the RV101 through CXR-L.

I looked at going around it on the glasses too because someone else talked about it in the Rokid's Discord. 
It doesn't work: UNPROCESSED and VOICE_RECOGNITION sources open fine but deliver silence for non-system apps, and the 4 mics come in as one backend, so the only source that actually produces audio is the beamformed one. The "Mic direction 360°" toggle in Hi Rokid only affects their translate scene, not what Nexus receives.

This is what GPT/Claude told me last time I checked , tldr: build yourself a custom AGC on your side. And to be honest, it's not really worth the hassle imo.
"What does work for the "person in front of you" case: a simple digital AGC on your side, per ~100 ms buffer, gain = min(target / rms, max) above the noise floor. In the field that took a speaker at 3 m from ~400 RMS to ~9000, plenty for ASR. Proper omni would need Rokid to expose it on the external interface. "

**Comment by nigelbreretonthelegend-sys** [comment]
Thanks — that's really useful and answers the scene-selection question.
One quick follow-up on the AGC test you mentioned. Was the ~3 m test where you increased the speaker from roughly 400 RMS to 9000 RMS performed on the same beamformed CXR-L PCM stream that Nexus exposes to plugins?
If so, that sounds potentially very useful for our use case. We don't necessarily need true omni or access to the individual microphones if we can make speech from the person in front of the wearer sufficiently intelligible for ASR by post-processing the Nexus PCM.
We're quite happy to test that experimentally rather than chase microphone-scene control that the RV101 doesn't expose through CXR-L.
If you still have the AGC parameters/code from your test — particularly target RMS, maximum gain, noise-floor threshold and whether you used any attack/release or gain smoothing — I'd be very interested in trying the same approach against our Nexus capture.
Thanks again — you've probably saved us quite a bit of unnecessary digging into the wrong SDK layer!

**Comment by Anezium** [comment]
Nope, not the same stream. 
That 400 → 9000 number came from another dev on the RV101, recording on the glasses themselves with AudioSource.MIC (the DSP beamformer). The Nexus stream goes through Rokid's other pipeline in cxrservice (a neural "SE" beamformer), which already applies a small AGC of its own, looks like +6 dB max from the logs. So both are wearer-focused, but I wouldn't reuse the numbers as-is, and I don't have their code either. I don't remember if you can have precise number with decompilation of the Hi Rokid Apk or check the CXR service while it runs.

What I'd start from on the Nexus PCM (16 kHz mono int16), then tune by ear (again it's made up by Fable 5.1, so I didn't really test it) :

- frame: 100 ms (1600 samples), compute RMS per frame
- noise floor: measure 1-2 s of silence first, threshold at ~2-3× that RMS; below it, gain = 1 (don't pump the room noise)
- target RMS: 3000-5000
- gain = min(target / rms, maxGain), maxGain around 20-30 (26-30 dB)
- smoothing: fast attack (~10-20 ms) so the first syllable isn't lost, slow release (~300-500 ms) so it doesn't breathe between words
- soft clip / limiter at the end, int16 overflows are ugly

The frames you get from the lease are ~20 ms each, so batch 5 of them per RMS window. If you get it working, I'd be curious to see the before/after RMS you measure on the Nexus stream, that would be the first real number for it.

**Comment by nigelbreretonthelegend-sys** [comment]
Hi Anezium,

Following your last note about the Nexus CXR stream, I now have actual
raw Nexus PCM and a controlled comparison which I thought might be
useful to you. There is probably too much detail for a sensible GitHub
issue comment.

I modified my existing Nexus feasibility plugin only to persist every
onAudioFrame ByteArray verbatim. The acquisition path itself was left
unchanged. Nexus reports 16 kHz, mono, PCM16LE. There is no gain,
filtering, denoise, resampling or compression in my code.

The controlled capture was clean: 7,490 frames / 4,793,600 bytes, with
persisted frame count, byte count and actual file size all matching.
Sequence 0–7489 had zero missing, duplicate or out-of-order frames and
zero backwards timestamps.

Test conditions

The distant source was an Ajeya Cotra podcast playing on a TV at
volume 10. I was square-on to the TV at 2.5 metres. A stationary phone
recording acted as a broad environmental reference, and Life Recorder
was running natively on the RV101 at the same time.

The ~150 s Nexus recording used five nominal 30 s states:

A — room baseline, TV off, wearer silent
B — TV speech only, wearer silent
C — TV speech + wearer speech
D — TV off + wearer speech
E — room baseline, wearer silent

Whole-state RMS:

                 Nexus     Life Recorder     Phone
A room             0.14          3.52          30.4
B TV only          0.89         15.82         644
C TV + wearer   2,114        4,041          2,760
D wearer only   1,989        3,466          2,585
E room             2.24          5.71          58.9

These are not calibrated SPL figures and Life Recorder is AAC while
Nexus is raw PCM, so I am not treating cross-recorder ratios as a
microphone specification. The within-path behaviour is the useful
part.

The Nexus result strongly supports what you told me about the
cxrservice wearer-focused SE path. In TV-only, the phone clearly
records the podcast while Nexus is effectively silent. As soon as I
speak, with the TV still running, Nexus rises into the thousands of
PCM counts. Turning the TV off while I continue speaking barely
changes the Nexus wearer level.

It also closes back down between my utterances while the TV continues.
It does not behave like a simple gate that opens for my voice and then
passes the room.

The most striking result is the exact-zero count in the raw Nexus PCM:

A room: 99.2%
B TV only: 70.7%
C TV + wearer: 7.1%
D wearer only: 16.7%
E room: 98.3%

So during rejected conditions cxrservice/Nexus is frequently
delivering literal digital zeros, not merely a conventionally
attenuated distant signal. I can't identify the exact upstream
algorithm from that, but it looks strongly nonlinear.

Life Recorder on the glasses is also wearer-focused but materially
less aggressive. Its short-time energy pattern aligns very closely
with Nexus (about 0.96 correlation in this test), particularly during
wearer speech, but the streams are definitely not identical. That
seems consistent with your point that Nexus uses the cxrservice neural
SE pipeline while on-glasses recording can use a different DSP path.

We also proved that Nexus and Life Recorder can capture concurrently:
the same real-world interval exists in both recordings with strong
acoustic alignment. That corrects an earlier suspicion I had that
acquiring the Nexus lease might necessarily stop Life Recorder.

AGC

I also tried the offline AGC idea on an earlier raw Nexus capture,
using roughly the parameters you suggested. It was informative, but
Test 2 shows why conventional AGC is probably the wrong tool for this
particular stream.

During known TV-only periods Nexus can have an RMS below 1 PCM count
and 70% exact zeros. Raising those periods towards a 3000–5000 RMS
target simply starts undoing useful rejection. In one confirmed
TV-only interval the raw Nexus RMS was about 6.7; the experimental AGC
raised it to about 64 while the phone reference showed the TV was
genuinely present.

My current conclusion is therefore to leave the suppression floor
alone and, if extra level is required for ASR, apply gain only after
detecting accepted wearer speech rather than run an always-active AGC.

For my application, Nexus now looks extremely good for wearer voice
control, commands, dictation and wearer recording. It looks
fundamentally unsuitable for environmental/life recording, and I'm not
yet convinced that enough distant/front-facing speech survives to
recover reliable transcripts even with post-processing.

I have retained the raw PCM, diagnostic log, simultaneous Life
Recorder and phone recordings, plus aligned plots and the full
analysis. If any of those would be useful for Nexus development, tell
me what you want and I'll make it available.

My next tests will compare Rokid's own native audio recorder and
video-recorder audio under the same 2.5 m / TV-volume-10 setup. They
won't use Nexus, so I won't clutter the Nexus thread with them unless
you think the results would add something to your understanding of the
RV101 microphone paths.

Thanks again for pointing me towards the distinction between the CXR
SE path and the on-glasses DSP path — having the raw PCM now makes
that distinction much easier to see experimentally.

Regards,

Nigel

**Comment by Anezium** [comment]
Damn, that's a great writeup, thanks for taking the time !
I knew the SE path was wearer-focused but I didn't know it was gating that hard. 
And yeah as you tested, the AGC was not a great tool for that..

So yes i guess I'll write it in the sdk/specs that the Mic in Nexus is only for the wearer voice.

And yes I'll gladly take the PCM + plots, that would be the first proper measurement of the Nexus stream anyone has done. A link here or a zip on the issue is fine. Also post the native recorder / video results too if you don't mind, they help understand the RV101 mic paths even if they don't touch Nexus.

Thanks again !

**Comment by nigelbreretonthelegend-sys** [comment]
I have a couple of reports ready for you. Contact me at nigelbrereton.thelegend@gmail.com and I'll attach them to a reply.
